File: ubuntu/CommunicationManager.py
import MessageHelper


class CommunicationManager:
    LISTENING = 1
    SENDING = 2
    STOPPED = 3

    def __init__(self, listener, sender):
        # In listening mode by default since client controls flow
        self.mode = self.STOPPED
        self.queued_msgs = []
        self.listener = listener
        self.sender = sender

    def listen(self):
        while self.mode == self.LISTENING:
            self.mode = self.listener.listen()

    def send(self):
        while self.mode == self.SENDING and self.queue_size() > 0:
            msg = self.dequeue()
            self.sender.send(msg)
        self.mode = self.LISTENING

    def done_sending(self):
        send_done_msg = MessageHelper.create_mode()
        self.sender.send(send_done_msg)

    def loop(self):
        while not self.mode == self.STOPPED:
            if self.mode == self.LISTENING:
                self.listen()
            elif self.mode == self.SENDING:
                self.send()
                self.done_sending()
            elif self.mode == self.STOPPED:
                break
            else:
                assert False
                #should not reach

    def enqueue(self, msg):
        self.queued_msgs.append(msg)

    def dequeue(self):
        return self.queued_msgs.pop(0)

    def queue_size(self):
        return len(self.queued_msgs)

File: android/src/org/jbull/jmessage/CommunicationManager.java
package org.jbull.jmessage;

import android.util.Log;

import java.io.IOException;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Timer;
import java.util.TimerTask;

public class CommunicationManager<T> {

    private final long sendPeriod;

    public Mode getMode() {
        return mode;
    }

    private Mode mode;
    private final Listener listener;
    private final Sender sender;

    private Queue<T> queue = new LinkedList<T>();


    public enum Mode {
        LISTENING,
        SENDING,
        STOPPED
    }


    /**
     * Instantiates a CommunicationManager that handles the sending and receiving looping as well as
     * switching modes between the two. It delegates the actual sending and listening to Sender and Listener objects
     * @param l a listener that will be called to listen to the connection and handle the received messages
     * @param s a sender that will be called to send to the connection
     * @param sendPeriodMs the time to remain in send mode before polling the server for messages
     */
    public CommunicationManager(Listener l, Sender<T> s, long sendPeriodMs) {
        this.listener = l;
        this.sender = s;
        this.mode = Mode.STOPPED;
        this.sendPeriod = sendPeriodMs;
    }

    public synchronized void listenLoop() throws IOException {
        Mode curMode = mode;
        while (curMode == Mode.LISTENING) {
            curMode = listener.listen();
        }
    }

    public synchronized void send(T msg) {
        queue.add(msg);
        if (mode == Mode.SENDING) {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        flushMessages();
                    } catch (IOException e) {
                        new RuntimeException("Error flushing messages in send", e);
                    }
                }
            }).start();
        }
    }

    public synchronized void flushMessages() throws IOException {
        while (mode == Mode.SENDING && !queue.isEmpty()) {
            sender.send(queue.remove());
        }
    }

    public void startSendTimer() {
        if (sendPeriod >= 0) {
            Timer timer = new Timer();
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    try {
                        switchMode(Mode.LISTENING);
                    } catch (IOException e) {
                        // TODO: handle better
                        e.printStackTrace();
                    }
                }
            }, sendPeriod);
        }
    }

    public void switchMode(Mode toMode) throws IOException {
        mode = toMode;
        if (toMode == Mode.SENDING) {
            startSendTimer();
            flushMessages();
        } else if (toMode == Mode.LISTENING) {
            //TODO this is not synchronized

            Message.Mode modeMessage = MessageHelper.createModeMessage(true, System.currentTimeMillis());
            sender.send(modeMessage);
            listenLoop();
        } else {
            assert toMode == Mode.STOPPED;
        }
    }

    public static interface Sender<T> {
        public void send(T msg) throws IOException;
    }

    public static interface Listener {
        public Mode listen() throws IOException;
    }
}
